rest/route/host.go

```diff
@@ -314,6 +316,133 @@ func getLimit(vals url.Values) (int, error) {
 	return limit, nil
 }
 
+////////////////////////////////////////////////////////////////////////
+//
+// POST /users/{user_id}/clear_hosts
+
+func makeClearHostsByUser(sc data.Connector, env evergreen.Environment) gimlet.RouteHandler {
+	return &clearHostsHandler{
+		sc:  sc,
+		env: env,
+	}
+}
+
+type clearHostsHandler struct {
+	user   string
+	dryRun bool
+
+	env evergreen.Environment
+	sc  data.Connector
+}
+
+func (ch clearHostsHandler) Factory() gimlet.RouteHandler {
+	return &clearHostsHandler{
+		sc:  ch.sc,
+		env: ch.env,
+	}
+}
+
+func (ch *clearHostsHandler) Parse(ctx context.Context, r *http.Request) error {
+	ch.user = gimlet.GetVars(r)["user_id"]
+	if ch.user == "" {
+		return gimlet.ErrorResponse{
+			Message:    "user cannot be empty",
+			StatusCode: http.StatusBadRequest,
+		}
+	}
+	vals := r.URL.Query()
+	ch.dryRun = vals.Get("dry_run") == "true"
+
+	return nil
+}
+
+func (ch *clearHostsHandler) Run(ctx context.Context) gimlet.Responder {
+	usr := gimlet.GetUser(ctx)
+	opts := model.APIHostParams{
+		UserSpawned: true,
+	}
+	// returns all up-hosts for user
+	hosts, err := ch.sc.FindHostsInRange(opts, ch.user)
+	if err != nil {
+		return gimlet.NewJSONErrorResponse(errors.Wrap(err, "database error getting hosts"))
+	}
+
+	volumes, err := ch.sc.FindVolumesByUser(ch.user)
+	if err != nil {
+		return gimlet.NewJSONErrorResponse(errors.Wrap(err, "database error getting volumes"))
+	}
+
+	toTerminate := model.APIClearHostsResults{
+		Hosts:   []string{},
+		Volumes: []string{},
+	}
+	currentTime := time.Now()
+
+	mgrCache := map[cloud.ManagerOpts]cloud.Manager{}
+	for _, v := range volumes {
+		if !v.NoExpiration && v.Expiration.Before(currentTime) { // already terminated
+			continue
+		}
+		toTerminate.Volumes = append(toTerminate.Volumes, v.ID)
+		if ch.dryRun {
+			continue
+		}
+		mgrOpts := cloud.ManagerOpts{
+			Provider: evergreen.ProviderNameEc2OnDemand,
+			Region:   cloud.AztoRegion(v.AvailabilityZone),
+		}
+		mgr, ok := mgrCache[mgrOpts]
+		if !ok {
+			mgr, err = cloud.GetManager(ctx, ch.env, mgrOpts)
+			if err != nil {
+				return gimlet.MakeJSONInternalErrorResponder(err)
+			}
+			mgrCache[mgrOpts] = mgr
+		}
+
+		// detach volume first
```

Comment:
why is it necessary to detach the volume before deleting it?

Reply:
AWS doesn't let you delete an attached volume. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-deleting-volume.html

Reply:
would it work to terminate the hosts and then the volumes?
(then the volumes would be free)

Reply:
sure, good idea